Metrics: Added Endpoint BPF compilation time.

eloycoto opened the pull request:
Things to consider in this PR: 

- Should `EndpointRegenerationTime` be an `histogramVec` instead of `CounterVec`?
- How do we document the labels on metrics? Is there any pattern in the doc or should I  make something?
- I added logfields.Subsys, I think that make sense, but I'm not sure is the correct one. Please share your thoughts
- I also added the endpointID label, not sure if we don't add because any reason, or it's ok to have in it.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/5329)
<!-- Reviewable:end -->

tgraf commented on pkg/endpoint/bpf.go:
```diff
@@ -638,9 +639,14 @@ func (e *Endpoint) regenerateBPF(owner Owner, epdir, reason string) (revnum uint
 			Debugf("BPF compilation completed")
+		metrics.EndpointRegenerationTime.With(map[string]string{
+			logfields.LogSubsys:  metrics.LabelEndpointRegenerationBPFCompilation,
+			logfields.EndpointID: e.StringID(),
```
The endpointID can be omitted. Prometheus labels should only describe static sets and not temporary or generated labels that have no meaning over time.

tgraf commented on pkg/metrics/metrics.go:
```diff
@@ -94,11 +103,11 @@ var (
 
 	// EndpointRegenerationTime is the total time taken to regenerate endpoint
-	EndpointRegenerationTime = prometheus.NewCounter(prometheus.CounterOpts{
+	EndpointRegenerationTime = prometheus.NewCounterVec(prometheus.CounterOpts{
```
This should be a histogram and not a counter

tgraf commented on pkg/endpoint/policy.go:
```diff
@@ -682,7 +682,10 @@ func (e *Endpoint) regenerate(owner Owner, reason string) (retErr error) {
 			regenerateTimeSec := float64(regenerateTimeNs) / float64(time.Second)
 			e.getLogger().WithField(logfields.EndpointRegenerationTime, time.Since(regenerateStart).String()).Info("Regeneration of endpoint has completed")
-			metrics.EndpointRegenerationTime.Add(regenerateTimeSec)
+			metrics.EndpointRegenerationTime.With(map[string]string{
```
This will be weird as if BPF compilation took place, we will account for the regeneration twice. Instead I think you have to:
1. Return a struct which contains the overall build time
2. Add labels depending on what operation was performed during regeneration:
   * BPF compilation
   * Policy evaluation
   * Proxy configuration
3. Have a single place where you update the metric and any combination of the above labels can be set.

That will allow us to see how long endpoint regenerations take and filter by label to see if only certain operations are subject to delays.

raybejjani commented on pkg/endpoint/metrics.go:
```diff
@@ -0,0 +1,53 @@
+)
+
+// regenerationMetric is a helper function to update metrics.EndpointRegenerationTime
+func regenerationMetric(scope string, duration time.Duration) {
```
Did you have other plans for this function? I would argue that leaving what this function does in `metricsEndpointRegeneration*` functions is actually more clear. In particular, if we have different labels for each then using this function becomes more complicated

raybejjani commented on pkg/endpoint/metrics.go:
```diff
@@ -0,0 +1,53 @@
+
+// metricsEndpointRegenerationBPF is a helper function to update
+// metrics.EndpointRegenerationTime with label BPF.
+func (e *Endpoint) metricsEndpointRegenerationBPF(duration time.Duration) {
```
Since these functions don't rely on endpoint, it might be nice to make them normal functions. Their dependencies would be a little clearer. It might also be useful to move them to the metrics package, but I'm less sure about that. If we did, we would be saying that the only way to update `EndpointRegenerationTime` is with these functions (we might need to make that metric private then). This might be a good thing.

In any case, I was a little unclear on what the function did until I read it. A verb in the function name might be more clear. `recordEndpointRegeneration*` but then we lose the metric bit.

raybejjani commented on pkg/endpoint/policy.go:
```diff
@@ -679,10 +679,9 @@ func (e *Endpoint) regenerate(owner Owner, reason string) (retErr error) {
-			metrics.EndpointRegenerationTime.Add(regenerateTimeSec)
-			metrics.EndpointRegenerationTimeSquare.Add(math.Pow(regenerateTimeSec, 2))
+			e.metricsEndpointRegenerationTotal(regenerateTimeNs)
+			metrics.EndpointRegenerationTimeSquare.Add(math.Pow(regenerateTimeNs.Seconds(), 2))
```
I think this metric is supposed to be tied closely to `EndpointRegenerationTime` so maybe update it as part of `metricsEndpointRegenerationTotal` instead? The same would then be true for `EndpointRegenerationCount` and any others

tgraf commented on pkg/metrics/metrics.go:
```diff
@@ -94,11 +98,11 @@ var (
-	EndpointRegenerationTime = prometheus.NewCounter(prometheus.CounterOpts{
+	EndpointRegenerationTime = prometheus.NewHistogramVec(prometheus.HistogramOpts{
 		Namespace: Namespace,
-		Name:      "endpoint_regeneration_seconds_total",
```
We can't just rename metrics as it will break dashboards. We need to add a new one and declare the old one to be obsoleted in 6 months or similar

tgraf reviewed (changes requested):
This looks great. It makes sense to wait for #5358 to be merged and then just expose all the measurements as done per that PR to be provided as metrics via labels.

joestringer commented on pkg/endpoint/metrics.go:
```diff
@@ -0,0 +1,56 @@
+// metricsEndpointRegenerationBPF is a helper function to update
+// metrics.EndpointRegenerationTime with label BPF.
+func metricsEndpointRegenerationBPF(duration time.Duration) {
+	regenerationMetric(labelEndpointRegenerationBPFCompilation, duration)
```
If it's only the scope that changes, it would be more lightweight to just declare the `map[string]string ...` as something like `compileScope` at the top of the package, then do:

`metrics.EndpointRegenerationTime.With(compileScope).Observe(...)` at each point.

eloycoto committed:
Metrics: Add endpoint regeneration metrics

Deprecated `endpoint_regeneration_seconds_total` due is a invalid
counter instead of Histogram

Added a new metric: `endpoint_regeneration_time_stats_seconds` that
pushes information with the following labels:
"bpfCompilation"
"mapSync"
"policyCalculation"
"prepareBuild"
"proxyConfiguration"
"proxyPolicyCalculation"
"proxyWaitForAck"
"waitingForCTClean"
"waitingForLock"

Also, added a new pkg/endoint/metrics.go where endpoints metrics are
send now.

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

ianvernon commented on pkg/endpoint/metrics.go:
```diff
@@ -0,0 +1,61 @@
+	prepareBuild           spanstat.SpanStat
+}
+
+// SendMetrics get the curret statistics and push the info to the given prometheus metrics.
```
"SendMetrics sends the regeneration statistics for this endpoint to Prometheus."

ianvernon commented on pkg/metrics/metrics.go:
```diff
@@ -71,6 +71,12 @@ var (
 	LabelDatapathFamily = "family"
 
+	// LabelScope is the label used to defined multiples scopes in the same
+	// metric.For example, one counter may measure a metric over the scope of
```
add a space after the period.

ianvernon commented on pkg/metrics/metrics.go:
```diff
@@ -93,11 +99,12 @@ var (
 	},
 		[]string{"outcome"})
 
+	// @TODO remove this metric in Cilium 1.4
```
Mark this as deprecated per the Go documentation conventions: https://blog.golang.org/godoc-documenting-go-code

ianvernon commented on pkg/endpoint/metrics.go:
```diff
@@ -0,0 +1,61 @@
+package endpoint
```
Add copyright header.

eloycoto replied:
Thanks for the review, done!

ianvernon commented on pkg/metrics/metrics.go:
```diff
@@ -93,11 +99,12 @@ var (
 	},
 		[]string{"outcome"})
 
+	// Deprecated: this metric will be removed on Cilium 1.4
```
"removed in"

ianvernon commented on pkg/endpoint/metrics.go:
```diff
@@ -0,0 +1,80 @@
+// Copyright 2016-2018 Authors of Cilium
```
Change to only be 2018.

joestringer commented on pkg/endpoint/metrics.go:
```diff
@@ -0,0 +1,80 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package metrics holds prometheus metrics objects and related utility functions. It
```
I think you mean that "this file holds..."? This is not the metrics package.

joestringer commented on pkg/metrics/metrics.go:
```diff
@@ -93,11 +99,12 @@ var (
 	},
 		[]string{"outcome"})
 
+	// Deprecated: this metric will be removed in Cilium 1.4
```
If we deprecate `endpoint_regeneration_square_seconds_total`, we should probably also deprecate `endpoint_regeneration_seconds_total`.

eloycoto replied:
Done!

joestringer reviewed (commented):
Some minor comments, mainly just that we can probably also deprecate the time_square of endpoint regeneration since it's being replaced with a native prometheus histogram type.
